File: windows/src-tauri/src/settings_store.rs
use anyhow::{Context, Result};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::path::PathBuf;
use std::sync::OnceLock;
use tokio::fs;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Project {
    pub path: String,
    pub name: Option<String>,
    pub github_filter: Option<String>,
    pub repo_root: Option<String>,
    /// Per-project prompt prefix. When set, overrides Settings.promptTemplate
    /// for tasks created against this project. Optional.
    #[serde(default)]
    pub prompt_template: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct GlobalViewSettings {
    #[serde(default)]
    pub excluded_paths: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GitHubSettings {
    #[serde(default = "default_gh_filter")]
    pub default_filter: String,
    #[serde(default = "default_poll_interval")]
    pub poll_interval_seconds: u64,
    #[serde(default = "default_merge_command")]
    pub merge_command: String,
}

fn default_gh_filter() -> String {
    "assignee:@me is:open".to_string()
}
fn default_poll_interval() -> u64 {
    60
}
fn default_merge_command() -> String {
    "gh pr merge ${number} --squash --delete-branch".to_string()
}

impl Default for GitHubSettings {
    fn default() -> Self {
        Self {
            default_filter: default_gh_filter(),
            poll_interval_seconds: default_poll_interval(),
            merge_command: default_merge_command(),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NotificationSettings {
    #[serde(default = "default_true")]
    pub notifications_enabled: bool,
    #[serde(default)]
    pub pushover_enabled: bool,
    pub pushover_token: Option<String>,
    pub pushover_user_key: Option<String>,
    #[serde(default)]
    pub render_markdown_image: bool,
}

fn default_true() -> bool {
    true
}

impl Default for NotificationSettings {
    fn default() -> Self {
        Self {
            notifications_enabled: true,
            pushover_enabled: false,
            pushover_token: None,
            pushover_user_key: None,
            render_markdown_image: false,
        }
    }
}

/// Byte-compatible with macOS `RemoteSettings` (Sources/.../SettingsStore.swift).
/// Field names — `host`, `remotePath`, `localPath`, `syncIgnores` — match exactly
/// so a settings.json moved between Mac and Windows keeps working.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct RemoteSettings {
    #[serde(default)]
    pub host: String,
    #[serde(default)]
    pub remote_path: String,
    #[serde(default)]
    pub local_path: String,
    /// nil = use mutagen::default_ignores()
    #[serde(default)]
    pub sync_ignores: Option<Vec<String>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SessionTimeoutSettings {
    #[serde(default = "default_timeout_minutes")]
    pub active_threshold_minutes: u64,
}

fn default_timeout_minutes() -> u64 {
    1440
}

impl Default for SessionTimeoutSettings {
    fn default() -> Self {
        Self {
            active_threshold_minutes: default_timeout_minutes(),
        }
    }
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub enum SelfCompactAction {
    QueuePrompt,
    CompactNow,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct SelfCompactRule {
    pub id: String,
    pub threshold_tokens: i64,
    pub action: SelfCompactAction,
    pub message: String,
}

impl SelfCompactRule {
    /// Byte-compatible defaults with macOS SelfCompactRule.defaults so a
    /// settings.json round-tripped between platforms keeps the same rule set.
    pub fn defaults() -> Vec<Self> {
        vec![
            Self {
                id: "ctx-500k".into(),
                threshold_tokens: 500_000,
                action: SelfCompactAction::QueuePrompt,
                message: "You are above the 500k context limit. Whenever it is convenient, use the kanban CLI to send yourself a self-compact.".into(),
            },
            Self {
                id: "ctx-600k".into(),
                threshold_tokens: 600_000,
                action: SelfCompactAction::QueuePrompt,
                message: "You are above the 600k context limit. Please compact yourself soon using the kanban CLI self-compact command.".into(),
            },
            Self {
                id: "ctx-700k".into(),
                threshold_tokens: 700_000,
                action: SelfCompactAction::QueuePrompt,
                message: "You are above the 700k context limit. Compact yourself IMMEDIATELY using the kanban CLI self-compact command.".into(),
            },
            Self {
                id: "ctx-750k".into(),
                threshold_tokens: 750_000,
                action: SelfCompactAction::CompactNow,
                message: "/compact".into(),
            },
        ]
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SelfCompactSettings {
    #[serde(default)]
    pub enabled: bool,
    #[serde(default = "default_self_compact_poll")]
    pub poll_interval_seconds: u64,
    #[serde(default = "SelfCompactRule::defaults")]
    pub rules: Vec<SelfCompactRule>,
}

fn default_self_compact_poll() -> u64 {
    30
}

impl Default for SelfCompactSettings {
    fn default() -> Self {
        Self {
            enabled: false,
            poll_interval_seconds: default_self_compact_poll(),
            rules: SelfCompactRule::defaults(),
        }
    }
}

/// Named API service binding for a coding assistant. Wraps the CLI with an
/// optional launcher prefix (e.g. `ollama launch`), a `--model` value, and
/// an optional base URL that becomes `ANTHROPIC_BASE_URL` (or the assistant's
/// equivalent) at launch time. Byte-compatible with macOS APIService.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct APIService {
    pub id: String,
    pub name: String,
    /// CodingAssistant raw value ("claude", "gemini", …). Stored as a string
    /// so an unknown id from a future macOS-built settings.json still parses.
    pub assistant: String,
    /// Shell command prepended before the assistant CLI. `None` = call the
    /// CLI directly. Example: `Some("ollama launch")`.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub launcher_prefix: Option<String>,
    /// Value passed to `--model`. `None` omits the flag entirely.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub model_flag: Option<String>,
    /// Base URL injected as an env var at launch (e.g. `ANTHROPIC_BASE_URL`).
    #[serde(default, skip_serializing_if = "Option::is_none", rename = "baseURL")]
    pub base_url: Option<String>,
}

impl APIService {
    /// Whether a `--` separator is required before the assistant's own flags.
    /// Mirrors macOS APIService.needsSeparator.
    pub fn needs_separator(&self) -> bool {
        self.launcher_prefix.is_some() || self.model_flag.is_some()
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct Settings {
    #[serde(default)]
    pub projects: Vec<Project>,
    #[serde(default)]
    pub global_view: GlobalViewSettings,
    #[serde(default)]
    pub github: GitHubSettings,
    #[serde(default)]
    pub notifications: NotificationSettings,
    #[serde(default)]
    pub session_timeout: SessionTimeoutSettings,
    #[serde(default)]
    pub prompt_template: String,
    #[serde(default = "default_issue_template")]
    pub github_issue_prompt_template: String,
    #[serde(default)]
    pub has_completed_onboarding: bool,
    /// Editor command (e.g. "code", "cursor", "nvim")
    #[serde(default)]
    pub editor: String,
    /// Terminal font size (8-24)
    #[serde(default = "default_terminal_font_size")]
    pub terminal_font_size: u32,
    /// Font size for the History tab transcript view (8-20). Matches the
    /// macOS `sessionDetailFontSize` AppStorage so a settings round-trip
    /// between platforms preserves the preference.
    #[serde(default = "default_session_detail_font_size")]
    pub session_detail_font_size: u32,
    #[serde(default)]
    pub remote: Option<RemoteSettings>,
    /// Automatic context-limit guard for Claude sessions. Byte-compatible
    /// with macOS Settings.selfCompact — a settings.json moved between Mac
    /// and Windows keeps its rule set.
    #[serde(default)]
    pub self_compact: SelfCompactSettings,
    /// Named API service bindings — see APIService for shape. Byte-compatible
    /// with macOS Settings.apiServices.
    #[serde(default)]
    pub api_services: Vec<APIService>,
    /// Maps `CodingAssistant.rawValue` → `APIService.id` for the per-assistant
    /// default service. Matches macOS Settings.defaultAPIServiceIds.
    #[serde(default)]
    pub default_api_service_ids: HashMap<String, String>,
}

impl Settings {
    /// Resolves the APIService to use for a card launch. Resolution order:
    /// per-card override → per-assistant default → none. Returns `None` when
    /// no binding exists or the referenced id no longer points at anything
    /// (e.g. the service was deleted but a card's override is stale).
    pub fn resolve_api_service(
        &self,
        card_override: Option<&str>,
        assistant_id: &str,
    ) -> Option<&APIService> {
        // Per-card override wins. Stale override → fall through to default
        // so deleting a service doesn't leave cards unlaunchable.
        if let Some(id) = card_override {
            if let Some(svc) = self.api_services.iter().find(|s| s.id == id) {
                return Some(svc);
            }
        }
        let default_id = self.default_api_service_ids.get(assistant_id)?;
        self.api_services.iter().find(|s| &s.id == default_id)
    }
}

fn default_terminal_font_size() -> u32 {
    15
}

fn default_session_detail_font_size() -> u32 {
    12
}

/// Which terminal runtime a card launches into. Picked per card on the gate
/// panel — there is no global default. Lives on `Link.card_runtime`; this
/// enum is shared so both `Link` and any future Settings code can refer to
/// the same canonical type. Serialized as the lowercase variant name so the
/// JSON values match the TS string union (`"windows"` / `"wsl"`).
#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "lowercase")]
pub enum CardRuntime {
    Windows,
    Wsl,
}

fn default_issue_template() -> String {
    "#${number}: ${title}\n\n${body}".to_string()
}

/// First time the settings file is read on this process, peek at the raw JSON
/// to see whether it came from the macOS app. macOS doesn't write the
/// Windows-only keys (`terminalFontSize`, `editor`, `cardRuntime`); when
/// they're absent serde's `#[serde(default)]` silently backfills the Windows
/// defaults. We use `terminalFontSize` alone as the sentinel — `cardRuntime`
/// has no default so its absence isn't macOS-specific.
///
/// No `schema_version` field is added — that would break the macOS byte-compat
/// invariant (sortedKeys+prettyPrinted JSON shared with the Swift app).
fn log_cross_platform_backfill_once(raw: &[u8]) {
    static LOGGED: OnceLock<()> = OnceLock::new();
    if LOGGED.get().is_some() {
        return;
    }
    let Ok(v) = serde_json::from_slice::<serde_json::Value>(raw) else { return };
    let obj = match v.as_object() {
        Some(o) => o,
        None => return,
    };
    if !obj.contains_key("terminalFontSize") {
        let _ = LOGGED.set(());
        crate::logging::info(
            "settings",
            "settings.json missing terminalFontSize — backfilling Windows defaults (file appears to originate from macOS)",
        );
    }
}

pub struct SettingsStore {
    file_path: PathBuf,
}

impl SettingsStore {
    pub fn new(base_path: Option<PathBuf>) -> Self {
        let base = base_path
            .unwrap_or_else(|| crate::coordination_store::kanban_data_dir());
        Self {
            file_path: base.join("settings.json"),
        }
    }

    pub async fn read(&self) -> Result<Settings> {
        if !self.file_path.exists() {
            let defaults = Settings::default();
            self.write(&defaults).await?;
            return Ok(defaults);
        }
        let data = fs::read(&self.file_path).await.context("read settings.json")?;
        log_cross_platform_backfill_once(&data);
        let settings: Settings = serde_json::from_slice(&data).unwrap_or_default();
        Ok(settings)
    }

    pub async fn write(&self, settings: &Settings) -> Result<()> {
        if let Some(parent) = self.file_path.parent() {
            fs::create_dir_all(parent).await.context("create settings dir")?;
        }
        let data = serde_json::to_vec_pretty(settings).context("serialize settings")?;
        let tmp = self.file_path.with_extension("json.tmp");
        fs::write(&tmp, &data).await.context("write settings tmp")?;
        fs::rename(&tmp, &self.file_path).await.context("rename settings tmp")?;
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn svc(id: &str, assistant: &str) -> APIService {
        APIService {
            id: id.into(),
            name: format!("svc-{id}"),
            assistant: assistant.into(),
            launcher_prefix: None,
            model_flag: None,
            base_url: None,
        }
    }

    #[test]
    fn resolve_prefers_card_override() {
        let mut s = Settings::default();
        s.api_services = vec![svc("a", "claude"), svc("b", "claude")];
        s.default_api_service_ids.insert("claude".into(), "a".into());
        let resolved = s.resolve_api_service(Some("b"), "claude").unwrap();
        assert_eq!(resolved.id, "b");
    }

    #[test]
    fn resolve_falls_back_to_per_assistant_default_when_no_override() {
        let mut s = Settings::default();
        s.api_services = vec![svc("a", "claude")];
        s.default_api_service_ids.insert("claude".into(), "a".into());
        let resolved = s.resolve_api_service(None, "claude").unwrap();
        assert_eq!(resolved.id, "a");
    }

    #[test]
    fn resolve_stale_override_falls_back_to_default_instead_of_failing() {
        // A card holds an override pointing at a deleted service. Without the
        // fallback the card would be unlaunchable; with it the per-assistant
        // default takes over. Matches macOS resolveAPIService.
        let mut s = Settings::default();
        s.api_services = vec![svc("kept", "claude")];
        s.default_api_service_ids.insert("claude".into(), "kept".into());
        let resolved = s.resolve_api_service(Some("deleted"), "claude").unwrap();
        assert_eq!(resolved.id, "kept");
    }

    #[test]
    fn resolve_returns_none_when_no_default_and_no_override() {
        let s = Settings::default();
        assert!(s.resolve_api_service(None, "claude").is_none());
    }

    #[test]
    fn settings_round_trip_preserves_api_services_and_defaults() {
        let mut s = Settings::default();
        s.api_services = vec![APIService {
            id: "ollama-1".into(),
            name: "Local Ollama".into(),
            assistant: "claude".into(),
            launcher_prefix: Some("ollama launch".into()),
            model_flag: Some("qwen3-coder-next:cloud".into()),
            base_url: Some("http://localhost:11434/v1".into()),
        }];
        s.default_api_service_ids.insert("claude".into(), "ollama-1".into());
        let json = serde_json::to_string(&s).unwrap();
        // baseURL key — explicit serde rename, mirrors macOS spelling.
        assert!(json.contains("\"baseURL\":"));
        let s2: Settings = serde_json::from_str(&json).unwrap();
        assert_eq!(s2.api_services.len(), 1);
        assert_eq!(s2.api_services[0].id, "ollama-1");
        assert_eq!(
            s2.default_api_service_ids.get("claude"),
            Some(&"ollama-1".to_string())
        );
        assert!(s2.api_services[0].needs_separator());
    }

    #[test]
    fn needs_separator_only_when_launcher_or_model_flag_set() {
        let bare = APIService {
            id: "bare".into(),
            name: "Bare".into(),
            assistant: "claude".into(),
            launcher_prefix: None,
            model_flag: None,
            base_url: Some("http://localhost".into()),
        };
        assert!(!bare.needs_separator(), "base_url alone never needs --");
    }
}

File: Sources/KanbanCodeCore/Infrastructure/SettingsStore.swift
import Foundation

/// Application settings, stored at ~/.kanban-code/settings.json.
public struct Settings: Codable, Sendable {
    public var projects: [Project]
    public var globalView: GlobalViewSettings
    public var github: GitHubSettings
    public var notifications: NotificationSettings
    public var remote: RemoteSettings?
    public var sessionTimeout: SessionTimeoutSettings
    public var promptTemplate: String
    public var githubIssuePromptTemplate: String
    public var columnOrder: [KanbanCodeColumn]
    public var hasCompletedOnboarding: Bool
    public var defaultAssistant: CodingAssistant?
    public var enabledAssistants: [CodingAssistant]
    /// User-defined API service configurations (e.g. Ollama, LiteLLM proxy).
    public var apiServices: [APIService]
    /// Maps `CodingAssistant.rawValue` → `APIService.id` for the default service per assistant.
    public var defaultAPIServiceIds: [String: String]
    /// Automatic context-limit guard for Claude sessions.
    public var selfCompact: SelfCompactSettings
    /// Global Codex board runtime and scheduler capacity.
    public var codexBoard: CodexBoardSettings

    public init(
        projects: [Project] = [],
        globalView: GlobalViewSettings = GlobalViewSettings(),
        github: GitHubSettings = GitHubSettings(),
        notifications: NotificationSettings = NotificationSettings(),
        remote: RemoteSettings? = nil,
        sessionTimeout: SessionTimeoutSettings = SessionTimeoutSettings(),
        promptTemplate: String = "",
        githubIssuePromptTemplate: String = "#${number}: ${title}\n\n${body}",
        columnOrder: [KanbanCodeColumn] = KanbanCodeColumn.allCases,
        hasCompletedOnboarding: Bool = false,
        defaultAssistant: CodingAssistant? = nil,
        enabledAssistants: [CodingAssistant] = CodingAssistant.allCases,
        apiServices: [APIService] = [],
        defaultAPIServiceIds: [String: String] = [:],
        selfCompact: SelfCompactSettings = SelfCompactSettings(),
        codexBoard: CodexBoardSettings = CodexBoardSettings()
    ) {
        self.projects = projects
        self.globalView = globalView
        self.github = github
        self.notifications = notifications
        self.remote = remote
        self.sessionTimeout = sessionTimeout
        self.promptTemplate = promptTemplate
        self.githubIssuePromptTemplate = githubIssuePromptTemplate
        self.columnOrder = columnOrder
        self.hasCompletedOnboarding = hasCompletedOnboarding
        self.defaultAssistant = defaultAssistant
        self.enabledAssistants = enabledAssistants
        self.apiServices = apiServices
        self.defaultAPIServiceIds = defaultAPIServiceIds
        self.selfCompact = selfCompact
        self.codexBoard = codexBoard
    }

    private enum CodingKeys: String, CodingKey {
        case projects, globalView, github, notifications, remote, sessionTimeout
        case promptTemplate, githubIssuePromptTemplate, columnOrder, hasCompletedOnboarding, defaultAssistant
        case enabledAssistants
        case apiServices, defaultAPIServiceIds
        case selfCompact, codexBoard
        case skill // backward-compat: old name for promptTemplate
    }

    // Backward-compatible decoding — new fields default gracefully.
    // IMPORTANT: every field is decoded independently via `try?`-style fallbacks
    // so a single bad value can never wipe out the whole settings object. A
    // real-world incident: a future version of the app wrote `codex` into
    // `enabledAssistants`, but the Swift enum only had claude/gemini — the
    // strict `[CodingAssistant]` decode threw, `BoardStore` caught it with
    // `try?`, and every project vanished from the UI on restart.
    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        projects = (try? container.decodeIfPresent([Project].self, forKey: .projects)) ?? []
        globalView = (try? container.decodeIfPresent(GlobalViewSettings.self, forKey: .globalView)) ?? GlobalViewSettings()
        github = (try? container.decodeIfPresent(GitHubSettings.self, forKey: .github)) ?? GitHubSettings()
        notifications = (try? container.decodeIfPresent(NotificationSettings.self, forKey: .notifications)) ?? NotificationSettings()
        remote = try? container.decodeIfPresent(RemoteSettings.self, forKey: .remote)
        sessionTimeout = (try? container.decodeIfPresent(SessionTimeoutSettings.self, forKey: .sessionTimeout)) ?? SessionTimeoutSettings()
        // Backward-compat: try "promptTemplate" first, fall back to "skill"
        promptTemplate = (try? container.decodeIfPresent(String.self, forKey: .promptTemplate))
            ?? (try? container.decodeIfPresent(String.self, forKey: .skill)) ?? ""
        githubIssuePromptTemplate = (try? container.decodeIfPresent(String.self, forKey: .githubIssuePromptTemplate))
            ?? "#${number}: ${title}\n\n${body}"
        columnOrder = (try? container.decodeIfPresent([KanbanCodeColumn].self, forKey: .columnOrder)) ?? KanbanCodeColumn.allCases
        hasCompletedOnboarding = (try? container.decodeIfPresent(Bool.self, forKey: .hasCompletedOnboarding)) ?? false
        defaultAssistant = try? container.decodeIfPresent(CodingAssistant.self, forKey: .defaultAssistant)
        // Assistants written by a newer version of the app (or a future CLI)
        // may include values this Swift enum doesn't know about. Decode loose
        // strings, then keep only the ones we recognize — future unknowns
        // round-trip as a no-op rather than nuking the whole config.
        if let rawAssistants = try? container.decodeIfPresent([String].self, forKey: .enabledAssistants) {
            let known = rawAssistants.compactMap(CodingAssistant.init(rawValue:))
            enabledAssistants = known.isEmpty ? CodingAssistant.allCases : known
        } else {
            enabledAssistants = CodingAssistant.allCases
        }
        apiServices = (try? container.decodeIfPresent([APIService].self, forKey: .apiServices)) ?? []
        defaultAPIServiceIds = (try? container.decodeIfPresent([String: String].self, forKey: .defaultAPIServiceIds)) ?? [:]
        selfCompact = (try? container.decodeIfPresent(SelfCompactSettings.self, forKey: .selfCompact)) ?? SelfCompactSettings()
        codexBoard = (try? container.decodeIfPresent(CodexBoardSettings.self, forKey: .codexBoard)) ?? CodexBoardSettings()
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(projects, forKey: .projects)
        try container.encode(globalView, forKey: .globalView)
        try container.encode(github, forKey: .github)
        try container.encode(notifications, forKey: .notifications)
        try container.encodeIfPresent(remote, forKey: .remote)
        try container.encode(sessionTimeout, forKey: .sessionTimeout)
        try container.encode(promptTemplate, forKey: .promptTemplate)
        try container.encode(githubIssuePromptTemplate, forKey: .githubIssuePromptTemplate)
        try container.encode(columnOrder, forKey: .columnOrder)
        try container.encode(hasCompletedOnboarding, forKey: .hasCompletedOnboarding)
        try container.encodeIfPresent(defaultAssistant, forKey: .defaultAssistant)
        try container.encode(enabledAssistants, forKey: .enabledAssistants)
        try container.encode(apiServices, forKey: .apiServices)
        try container.encode(defaultAPIServiceIds, forKey: .defaultAPIServiceIds)
        try container.encode(selfCompact, forKey: .selfCompact)
        try container.encode(codexBoard, forKey: .codexBoard)
        // Note: "skill" is NOT encoded — only read for backward-compat
    }
}

public struct CodexBoardSettings: Codable, Sendable, Equatable {
    public static let concurrencyRange = 1 ... 32

    public var runtime: CodexRuntimeBackend
    private var storedMaxConcurrency: Int

    public var maxConcurrency: Int {
        get { storedMaxConcurrency }
        set { storedMaxConcurrency = Self.clampedConcurrency(newValue) }
    }

    public init(runtime: CodexRuntimeBackend = .cliTmux, maxConcurrency: Int = 3) {
        self.runtime = runtime
        storedMaxConcurrency = Self.clampedConcurrency(maxConcurrency)
    }

    private enum CodingKeys: String, CodingKey {
        case runtime, maxConcurrency
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        runtime = (try? container.decodeIfPresent(CodexRuntimeBackend.self, forKey: .runtime)) ?? .cliTmux
        let concurrency = (try? container.decodeIfPresent(Int.self, forKey: .maxConcurrency)) ?? 3
        storedMaxConcurrency = Self.clampedConcurrency(concurrency)
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(runtime, forKey: .runtime)
        try container.encode(maxConcurrency, forKey: .maxConcurrency)
    }

    public static func clampedConcurrency(_ value: Int) -> Int {
        min(max(value, concurrencyRange.lowerBound), concurrencyRange.upperBound)
    }
}

public struct GlobalViewSettings: Codable, Sendable {
    public var excludedPaths: [String]

    public init(excludedPaths: [String] = []) {
        self.excludedPaths = excludedPaths
    }
}

public struct GitHubSettings: Codable, Sendable {
    public var defaultFilter: String
    public var pollIntervalSeconds: Int
    public var mergeCommand: String

    public static let defaultMergeCommand = "gh pr merge ${number} --squash --delete-branch"

    public init(defaultFilter: String = "assignee:@me is:open", pollIntervalSeconds: Int = 60, mergeCommand: String? = nil) {
        self.defaultFilter = defaultFilter
        self.pollIntervalSeconds = pollIntervalSeconds
        self.mergeCommand = mergeCommand ?? Self.defaultMergeCommand
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        defaultFilter = try c.decodeIfPresent(String.self, forKey: .defaultFilter) ?? "assignee:@me is:open"
        pollIntervalSeconds = try c.decodeIfPresent(Int.self, forKey: .pollIntervalSeconds) ?? 60
        mergeCommand = try c.decodeIfPresent(String.self, forKey: .mergeCommand) ?? Self.defaultMergeCommand
    }
}

public enum PushoverMode: String, Codable, Sendable, CaseIterable {
    case disabled
    case enabled
    case whenLidClosed
}

public struct NotificationSettings: Codable, Sendable {
    public var pushoverMode: PushoverMode
    public var pushoverToken: String?
    public var pushoverUserKey: String?
    public var renderMarkdownImage: Bool

    /// Backward-compatible convenience: true when pushover should be configured at all.
    public var pushoverEnabled: Bool { pushoverMode != .disabled }

    public init(pushoverMode: PushoverMode = .disabled, pushoverToken: String? = nil, pushoverUserKey: String? = nil, renderMarkdownImage: Bool = false) {
        self.pushoverMode = pushoverMode
        self.pushoverToken = pushoverToken
        self.pushoverUserKey = pushoverUserKey
        self.renderMarkdownImage = renderMarkdownImage
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        // Backward compat: read old Bool pushoverEnabled if pushoverMode is missing
        if let mode = try c.decodeIfPresent(PushoverMode.self, forKey: .pushoverMode) {
            pushoverMode = mode
        } else {
            let legacy = try c.decodeIfPresent(Bool.self, forKey: .pushoverEnabled) ?? false
            pushoverMode = legacy ? .enabled : .disabled
        }
        pushoverToken = try c.decodeIfPresent(String.self, forKey: .pushoverToken)
        pushoverUserKey = try c.decodeIfPresent(String.self, forKey: .pushoverUserKey)
        renderMarkdownImage = try c.decodeIfPresent(Bool.self, forKey: .renderMarkdownImage) ?? false
    }

    private enum CodingKeys: String, CodingKey {
        case pushoverMode, pushoverEnabled, pushoverToken, pushoverUserKey, renderMarkdownImage
    }

    public func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        try c.encode(pushoverMode, forKey: .pushoverMode)
        try c.encodeIfPresent(pushoverToken, forKey: .pushoverToken)
        try c.encodeIfPresent(pushoverUserKey, forKey: .pushoverUserKey)
        try c.encode(renderMarkdownImage, forKey: .renderMarkdownImage)
    }
}

public struct RemoteSettings: Codable, Sendable, Equatable {
    public var host: String
    public var remotePath: String
    public var localPath: String
    public var syncIgnores: [String]?  // nil = use MutagenAdapter.defaultIgnores

    public init(host: String, remotePath: String, localPath: String, syncIgnores: [String]? = nil) {
        self.host = host
        self.remotePath = remotePath
        self.localPath = localPath
        self.syncIgnores = syncIgnores
    }
}

public struct SessionTimeoutSettings: Codable, Sendable {
    public var activeThresholdMinutes: Int

    public init(activeThresholdMinutes: Int = 1440) {
        self.activeThresholdMinutes = activeThresholdMinutes
    }
}

public enum SelfCompactAction: String, Codable, Sendable, CaseIterable {
    case queuePrompt
    case compactNow

    public var displayName: String {
        switch self {
        case .queuePrompt: "Queue prompt"
        case .compactNow: "Compact now"
        }
    }
}

public struct SelfCompactRule: Identifiable, Codable, Sendable, Equatable {
    public var id: String
    public var thresholdTokens: Int
    public var action: SelfCompactAction
    public var message: String

    public init(
        id: String,
        thresholdTokens: Int,
        action: SelfCompactAction,
        message: String
    ) {
        self.id = id
        self.thresholdTokens = thresholdTokens
        self.action = action
        self.message = message
    }

    public static let defaults: [SelfCompactRule] = [
        SelfCompactRule(
            id: "ctx-500k",
            thresholdTokens: 500_000,
            action: .queuePrompt,
            message: "You are above the 500k context limit. Whenever it is convenient, use the kanban CLI to send yourself a self-compact."
        ),
        SelfCompactRule(
            id: "ctx-600k",
            thresholdTokens: 600_000,
            action: .queuePrompt,
            message: "You are above the 600k context limit. Please compact yourself soon using the kanban CLI self-compact command."
        ),
        SelfCompactRule(
            id: "ctx-700k",
            thresholdTokens: 700_000,
            action: .queuePrompt,
            message: "You are above the 700k context limit. Compact yourself IMMEDIATELY using the kanban CLI self-compact command."
        ),
        SelfCompactRule(
            id: "ctx-750k",
            thresholdTokens: 750_000,
            action: .compactNow,
            message: "/compact"
        ),
    ]
}

public struct SelfCompactSettings: Codable, Sendable, Equatable {
    public var enabled: Bool
    public var pollIntervalSeconds: Int
    public var rules: [SelfCompactRule]

    public init(
        enabled: Bool = false,
        pollIntervalSeconds: Int = 30,
        rules: [SelfCompactRule] = SelfCompactRule.defaults
    ) {
        self.enabled = enabled
        self.pollIntervalSeconds = pollIntervalSeconds
        self.rules = rules
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        enabled = try c.decodeIfPresent(Bool.self, forKey: .enabled) ?? false
        pollIntervalSeconds = try c.decodeIfPresent(Int.self, forKey: .pollIntervalSeconds) ?? 30
        let decodedRules = (try? c.decodeIfPresent([SelfCompactRule].self, forKey: .rules)) ?? SelfCompactRule.defaults
        rules = decodedRules.isEmpty ? SelfCompactRule.defaults : decodedRules
    }

    private enum CodingKeys: String, CodingKey {
        case enabled, pollIntervalSeconds, rules
    }
}

/// Reads and writes ~/.kanban-code/settings.json.
/// Caches settings in memory and only re-reads from disk when mtime changes.
public actor SettingsStore {
    private let filePath: String
    private let encoder: JSONEncoder
    private let decoder: JSONDecoder
    private var cachedSettings: Settings?
    private var cachedMtime: Date?

    public init(basePath: String? = nil) {
        let base = basePath ?? (NSHomeDirectory() as NSString).appendingPathComponent(".kanban-code")
        self.filePath = (base as NSString).appendingPathComponent("settings.json")

        let encoder = JSONEncoder()
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        self.encoder = encoder

        self.decoder = JSONDecoder()
    }

    /// Invalidate the in-memory cache so the next read() re-reads from disk.
    public func invalidateCache() {
        cachedSettings = nil
        cachedMtime = nil
    }

    /// Read settings, creating defaults if file doesn't exist.
    /// Returns cached value if the file hasn't changed since last read.
    public func read() throws -> Settings {
        let fileManager = FileManager.default
        guard fileManager.fileExists(atPath: filePath) else {
            let defaults = Settings()
            try write(defaults)
            return defaults
        }

        // Check mtime — return cached if unchanged
        let attrs = try? fileManager.attributesOfItem(atPath: filePath)
        let mtime = attrs?[.modificationDate] as? Date
        if let cached = cachedSettings, let cachedMtime, mtime == cachedMtime {
            return cached
        }

        let data = try Data(contentsOf: URL(fileURLWithPath: filePath))
        let settings = try decoder.decode(Settings.self, from: data)
        cachedSettings = settings
        cachedMtime = mtime
        return settings
    }

    /// Write settings atomically.
    public func write(_ settings: Settings) throws {
        let fileManager = FileManager.default
        let dir = (filePath as NSString).deletingLastPathComponent
        try fileManager.createDirectory(atPath: dir, withIntermediateDirectories: true)

        let data = try encoder.encode(settings)
        let tmpPath = filePath + ".tmp"
        try data.write(to: URL(fileURLWithPath: tmpPath))
        _ = try? fileManager.removeItem(atPath: filePath)
        try fileManager.moveItem(atPath: tmpPath, toPath: filePath)

        // Update cache with the just-written value
        cachedSettings = settings
        cachedMtime = (try? fileManager.attributesOfItem(atPath: filePath))?[.modificationDate] as? Date
    }

    /// The file path for external access.
    public var path: String { filePath }

    // MARK: - Project convenience methods

    /// Add a project to settings. Throws if path already exists.
    public func addProject(_ project: Project) throws {
        var settings = try read()
        guard !settings.projects.contains(where: { $0.path == project.path }) else {
            throw SettingsError.duplicateProject(project.path)
        }
        settings.projects.append(project)
        try write(settings)
    }

    /// Update an existing project (matched by path).
    public func updateProject(_ project: Project) throws {
        var settings = try read()
        guard let index = settings.projects.firstIndex(where: { $0.path == project.path }) else {
            throw SettingsError.projectNotFound(project.path)
        }
        settings.projects[index] = project
        try write(settings)
    }

    /// Remove a project by path.
    public func removeProject(path: String) throws {
        var settings = try read()
        guard settings.projects.contains(where: { $0.path == path }) else {
            throw SettingsError.projectNotFound(path)
        }
        settings.projects.removeAll { $0.path == path }
        try write(settings)
    }

    /// Save the reordered projects list.
    public func reorderProjects(_ projects: [Project]) throws {
        var settings = try read()
        settings.projects = projects
        try write(settings)
    }
}

public enum SettingsError: LocalizedError {
    case duplicateProject(String)
    case projectNotFound(String)

    public var errorDescription: String? {
        switch self {
        case .duplicateProject(let path): "Project already configured: \(path)"
        case .projectNotFound(let path): "Project not found: \(path)"
        }
    }
}
